SimpleFactoring.sol

```diff
@@ -7,27 +7,36 @@ pragma solidity >=0.7.0 <0.9.0;
  * @dev Implements simple factoring
  */
 contract SimpleFactoring {
-    address private boss; // Owner of contract
-    uint8 public commision = 10; // 10% commision on invoice sales;
+    address payable private boss; // Owner of contract
+    uint8 public commision = 2; // 2% commision on invoice sales;
 
     struct Invoice {
+        uint256 index; // Index of Invoice = position in array
         uint256 dueDate; // Deadline of payment
         address payer; // Customer
         bool settled; // If true, the invoice has been paid
-        uint256 total; // Amount to be paid in Ether
-        uint256 resellValue; // Invoice price in case of a sale
-        uint8 resellCount; // Number of resells
+        uint256 total; // Amount to be paid in Wei
+        uint256 resellPrice; // Invoice price in case of a sale
     }
 
     mapping(address => Invoice[]) private invoices;
+    uint256 private userCount;
+    uint256 private invoiceCount;
+    mapping (uint256 => address) private users;
 
     struct Offer {
+        uint256 index; // Index of Offer = position in array
         Invoice invoice; // Invoice to be sold
-        address payable payee; // Seller
+        address payable seller; // Seller
     }
 
     Offer[] private offers;
 
+    struct PayableInvoice {
+        Invoice invoice; // Invoice to be sold
+        address payable beneficiary; // Beneficiary of the invoice
+    }
+
     // Modifier to check if caller is owner
     modifier bossGuard {
```

Comment:
Szerintem itt nem azt kellene ellenőrizni, hogy a contract owner-e a metódus hívója, hanem azt, hogy a ki a tulaja. Elképzelhető, hogy ezen egy fokot bonyolítani is kell majd, mert van olyan művelet amit csak a payee hajthat végre és van olyan is amit csak a payer.

Reply:
A bossGuard-ot azért hoztam létre, hogy az owner elérje az "admin" műveleteket, más viszont nem.  Ha jól értem, te nem ezt módosítanád, hanem új guardokat hoznál létre a Payee / Payer szerepekhez.
Milyen konkrét esetekre gondolsz? Mert a jelenlegi implementációban én nem látom hol szükségesek ezek. A jogosultságok kezelése általában msg.sender-rel vannak biztosítva, pl. Invoice műveleteknél biztosan csak a Payee fér hozzá a sajátjaihoz.